### PR TITLE
Attach user metadata from commands to corresponding history events

### PR DESCRIPTION
**Pull request by gow**
## What changed?
In this PR we are attaching the user metadata if found in the command to the corresponding history events that are being created. 

## Why?
Users need a way to set arbitrary information on workflows and events. This would be later used in the clients to build custom tooling (ex: UI showing additional information about a timer event etc)

## How did you test it?
Added unit tests

## Potential risks
N/A. It is backward compatible. 

## Documentation
N/A

## Is hotfix candidate?
No

### CODE REVIEW

File: service/history/workflow_task_handler.go [hunk 1]
```diff
@@ -345,6 +349,15 @@ func (handler *workflowTaskHandlerImpl) handleCommand(
 		}
```
Comment (gow) [inline service/history/workflow_task_handler.go:349]: Looks like we have a command registery as well where we dynamically register and handle commands on line 337 above. `ch, ok := handler.commandHandlerRegistry.Handler(command.GetCommandType())`

There are currently just 2 commands being registered dynamically (`COMMAND_TYPE_SCHEDULE_NEXUS_OPERATION` & `COMMAND_TYPE_REQUEST_CANCEL_NEXUS_OPERATION`). Both of these create history events. Do we care about user metadata in these commands (or any that may be registered in the future). cc @yiminc @cretz

Comment (gow) [reply service/history/workflow_task_handler.go:349]: @bergundy thoughts?

Comment (cretz) [reply service/history/workflow_task_handler.go:349]: Unless it's a real problem, I think we should consistently copy all command metadata to all related event metadata regardless of the internal code paths.

**Comment by cretz** [comment]
Will let others more familiar with this repo review. But I want to confirm that some features will be present:

* Is the user metadata on `StartWorkflowExecution` and `SignalWithStartWorkflowExecutionRequest` copied to the start event?
* Is the user metadata on `NewWorkflowExecutionInfo` copied to the start event on schedules?
* Is the user metadata on `WorkflowExecutionConfig` returned from describe populated with the metadata the workflow was started with?

**Comment by gow** [comment]
> Will let others more familiar with this repo review. But I want to confirm that some features will be present:
> 
> * Is the user metadata on `StartWorkflowExecution` and `SignalWithStartWorkflowExecutionRequest` copied to the start event?
> * Is the user metadata on the child workflow command copied to the workflow start event?
> * Is the user metadata on `NewWorkflowExecutionInfo` copied to the start event on schedules?
> * Is the user metadata on `WorkflowExecutionConfig` returned from describe populated with the metadata the workflow was started with?

This is PR is just for commands only (to keep PRs short). I'm working on another PR to include user metadata in all the other areas you mentioned.

File: service/history/workflow_task_handler.go [hunk 2]
```diff
@@ -345,6 +349,15 @@ func (handler *workflowTaskHandlerImpl) handleCommand(
...
 		return nil, err
 	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	if historyEvent != nil && command.UserMetadata != nil {
```
Comment (yycptt) [inline service/history/workflow_task_handler.go:357]: As synced offline I think ideally the returned historyEvent should not be modified and instead the underlying mutable state and history build can provide a method for mutating added history events. But we can do that later I guess. 
Please add some TODO here though.

File: service/history/workflow_task_handler.go [hunk 3]
```diff
@@ -345,6 +349,15 @@ func (handler *workflowTaskHandlerImpl) handleCommand(
...
+		historyEvent.UserMetadata = command.UserMetadata
```
Comment (yycptt) [inline service/history/workflow_task_handler.go:358]: I get the idea that UserMetadata applies to all history events so it's a field on the Command type not on individual command attributes. But my understanding is ^ is not the case for history events?

**Review by yycptt — APPROVED** [review]
Workflow task handler logic LGTM

File: service/history/workflow_task_handler.go (hunk 2 above)
Comment (gow) [reply service/history/workflow_task_handler.go:357]: Will do.

Comment (gow) [reply service/history/workflow_task_handler.go:359]: Done.

File: service/history/workflow_task_handler.go (hunk 3 above)
Comment (gow) [reply service/history/workflow_task_handler.go:360]: Synced offline. This is to keep the server changes minimum.

File: service/history/workflow_task_handler.go (hunk 1 above)
Comment (gow) [reply service/history/workflow_task_handler.go]: This would require a larger refactor. Can we do it if actually required?

**Pull request merged**